gpu-screen-recorder{,-gtk}: Update to unstable-2024-07-05 (+ systemd unit and capabilities)

timschumi opened the pull request:
## Description of changes

Non-NixOS related changes can be found [here](https://git.dec05eba.com/gpu-screen-recorder-gtk/tree/com.dec05eba.gpu_screen_recorder.appdata.xml). Relevant changes start from 2024-05-21 for `gpu-screen-recorder` and from 3.7.6 for `gpu-screen-recorder-gtk`.

NixOS-specific changes include the following:
- Switch from the script-based build system to meson
- Install the upstream-provided systemd user service file
- Create `programs.gpu-screen-recorder` to host an option for generating `setcap` wrappers
- Return `gpu-screen-recorder-gtk` to using tarballs instead of Git (as the developer explicitly asked for that in #268443)

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [x] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

timschumi committed:
gpu-screen-recorder: unstable-2024-05-21 -> unstable-2024-07-05

timschumi committed:
gpu-screen-recorder: Enable installing systemd user units

timschumi commented on pkgs/applications/video/gpu-screen-recorder/default.nix:
```diff
@@ -37,16 +43,21 @@ stdenv.mkDerivation {
+    # FIXME: systemd unit file gets installed to absolute path.
+    "-Dsystemd=false"
+
+    # FIXME: Capabilities can not be set if not running the build with root permissions.
+    "-Dcapabilities=false"
```
Would appreciate to have some feedback/help on this, I'm unfortunately too little of a NixOS aficionado to figure this out myself (or if it is even possible).

```
Running custom install script '/build/extra/meson_post_install.sh'
unable to set CAP_SETFCAP effective capability: Operation not permitted
\n!!! Please re-run install as root\n
unable to set CAP_SETFCAP effective capability: Operation not permitted
FAILED: install script '/build/extra/meson_post_install.sh' failed with exit code 1.
```

dec05eba replied:
In the case of nix it has to be done differently: https://github.com/NixOS/nixpkgs/issues/275481

timschumi replied:
I'll take a look, thanks for the pointer.

timschumi replied:
Implemented now and seems to work great.

@sersorrel I assigned you a co-author attribution on the respective commit since this is based on your issue report, please let me know in case you want me to omit it or something is wrong with the metadata.

dec05eba commented:
https://git.dec05eba.com/gpu-screen-recorder-gtk/tree/com.dec05eba.gpu_screen_recorder.appdata.xml there is a changelog for the larger changes (this is the changelog used by the flatpak)

timschumi committed:
maintainers: add timschumi

timschumi committed:
gpu-screen-recorder: Add program module for creating setcap wrappers

Co-authored-by: oddlama <oddlama@oddlama.org>
Co-authored-by: ash <ash@sorrel.sh>

oddlama reviewed (approved):
Can confirm that root wrappers work as expected, thanks!

timschumi committed:
gpu-screen-recorder-gtk: 3.7.6 -> unstable-2024-07-05

timschumi commented:
Changes since last push:

- Rebased onto current `master` to resolve `gpu-screen-recorder-gtk`-related conflicts.
- Update `gpu-screen-recorder-gtk` in lockstep. This includes returning to an un-"tagged" revision to match `gpu-screen-recorder`.
- Switch `gpu-screen-recorder-gtk` to the Meson build system (direct result of previous entry).
- Return to tarball downloading for `gpu-screen-recorder-gtk` as originally requested by the maintainer in #268443.
- Uprev from 2024-07-04 to 2024-07-05.

oddlama reviewed (approved):
LGTM

keenanweaver commented:
Looks great. Thanks for making this! I've been meaning to make something for this but never got around to it.

Could the work well with the [unstable updater](https://github.com/NixOS/nixpkgs/blob/master/pkgs/common-updater/unstable-updater.nix)? gpu-screen-recorder in the past has been lacking updates in nixpkgs.

timschumi commented:
> Could this work well with the [unstable updater](https://github.com/NixOS/nixpkgs/blob/master/pkgs/common-updater/unstable-updater.nix)? gpu-screen-recorder in the past has been lacking updates in nixpkgs.

It doesn't appear to be compatible without any changes, and given that we ideally want to keep using snapshot tarballs and since there are no usable tags anyways (which appears to account for roughly 50% of unstable updater code), it's most likely easier to just script something separate.

SuperSandro2000 commented on pkgs/applications/video/gpu-screen-recorder/gpu-screen-recorder-gtk.nix:
```diff
@@ -1,51 +1,59 @@
-    url = "https://repo.dec05eba.com/gpu-screen-recorder-gtk";
-    rev = "cd777c1506e20514df4b97345e480051cbaf9693";
-    hash = "sha256-ZBYYsW75tq8TaZp0F4v7YIHKHk/DFBIGy3X781ho2oE=";
+  # printf "r%s.%s\n" "$(git rev-list --count HEAD)" "$(git rev-parse --short HEAD)"
```
We should add a really short comment for what this is good for.

SuperSandro2000 commented on nixos/modules/programs/gpu-screen-recorder.nix:
```diff
@@ -0,0 +1,43 @@
+        type = lib.types.bool;
+        default = false;
+        description = ''
+          Generate setcap wrappers.
```
```suggestion
          Wether to generate setcap wrappers.
```

SuperSandro2000 commented on nixos/modules/programs/gpu-screen-recorder.nix:
```diff
@@ -0,0 +1,43 @@
+
+      wrapCapabilities = lib.mkOption {
+        type = lib.types.bool;
+        default = false;
```
Why not default this to true? Also does this even need to be an option?

timschumi replied:
I wouldn't dare to impose automatically generating these wrappers on anyone who is just loading the program module definition and doesn't actually use gpu-screen-recorder (do note that we don't have an `enable` option here, as it would be unclear whether it just installs the package or actually starts the included background service).

Unless you are asking why I haven't put the "generating wrappers" part into `pkgs/...`, in which case the answer would be "I couldn't figure out how".

keenanweaver replied:
I'm in favor of setting it enabled as default. It's a big boon of nix modules to 'take care' of these sorts of options. Getting prompted for root when starting a replay is kinda annoying, and when it's resolved by just setting `programs.gpu-screen-recorder.enable` to true, that's really nice. If people just want the binaries, they'd just add the package names to their systemPackages. Maybe you can set a note in the description what it does and any security implication it may have?

oddlama replied:
I think the issue here is that changing this setting enabled-by-default will cause every system in the world to create these wrappers and depend on gpu-screen-recorder, because there is no option like `programs.gpu-screen-recorder.enable`. So this would be an absolute no-go.

oddlama replied:
I guess we could rename this option to `programs.gpu-screen-recorder.enable` if that's more consistent, but you'd still have to set an option either way.

timschumi replied:
Updated the option name to `.enable` (and updated the option description accordingly), if flashrom does this then we can't be too far off.
